README.md
# IMDb PHP API

## How To Use

### Create an instance

    // IMDb ( [bool $anonymise = false [, bool $summary = true [, int $titlesLimit = 0]]] )
    $imdb = new IMDb(true, true, 0);	// anonymise requests to prevent IP address getting banned, summarise returned data, unlimited films returned


### Search for a movie by title

    // Returns an array containing objects of matching titles
    $movies = $imdb->find_by_title("The Godfather"); 
    
    // $movies[0]->title => "The Godfather"
    // $movies[0]->year => "1972"
    // $movies[0]->tconst => "tt0068646"
    // ... 
    // $movies[1]->title => "The Godfather: Part II"
    // ...

### Get a movie by its imdb_id

    // Returns an object containing the movie's data
    $movie = $imdb->find_by_id("tt0068646");

    // $movie->title => "The Godfather"
    // $movie->rating => "8.1"
    // $movie->year => "1972"

### Search for a person by name

    // Returns an array containing objects of matching names
    $people = $imdb->person_by_name("Christian Bale"); 
    
    // $people[0]->name => "Christian Bale"
    // $people[0]->nconst => "nm0000288"
    // ... 

### Get a person by their imdb_id

    // Returns an object containing the person's data
    $person = $imdb->person_by_id("nm0000288");

    // $person->name => "Christian Bale"
    // $person->birthday_datestamp => "1974-01-30"
    // $person->birthplace => "Haverfordwest, Pembrokeshire, Wales, UK"
    
### Charts

    // Returns an array containing objects of of the top 250 films
    $movies = $imdb->chart_top();
    // Or for the Bottom 100...
    $movies = $imdb->chart_bottom();
    
    // $movies[0]->title => "The Shawshank Redemption"
    // $movies[0]->year => "1994"
    // $movies[0]->tconst => "tt0111161"
    // ... 
    // $movies[1]->title => "The Godfather"
    // ...
    
    // Similarly, the following function will return the Box Office (US) respectively.
    // Note that the format of the data returned differs slightly from the above 2 charts
    $movies = $imdb->boxoffice();
    

CHANGES.md
# Change Log

## v0.4.1 - June 22, 2012
* Fixed: build_url() now generates a signature to sign the request, this fixes find_by_title() returning an empty result set.
* Known Issue: Anonymising find_by_title() requests will always return an empty result set. As of writing this, the only solution is simply to set $anonymise to false.

## v0.4 - March 9, 2012
* Added: Can now get top charts data including the Top 250, Bottom 100, and Box Office (US)

## v0.3 - March 8, 2012
* Added: Can now search for people by name (or a person with their IMDb ID) (test files included)
* Added: Film tagline now returned.
* Added: Arrays of genres, writers, directors and actors are now returned in addition to a comma-separated list of them.
* Fixed: PHP warning when film has no genres.
* Fixed: Adult genre filter now checks all genres instead of only the first one listed.

## v0.2.3 - December 11, 2011
* Added runtime (in minutes) to returned data from find_by_id().

## v0.2.2 - December 11, 2011
* find_by_title() now accepts optional 'year' parameter to limit results to titles from a given year

## v0.2.1 - December 11, 2011
* Option to ignore certain types of film, by default, tv_series, tv_special and video_game types are ignored
* Option to ignore Adult genre or X rated films - this is turned on by default
* Option to override above settings and force a title to be returned with find_by_id()

## v0.2.0 - December 11, 2011
* Summarised data now returned by default (as opposed to literally all information, which is not needed by most people)
* Basic error handling - "response" key added to summarised data, 0 indicates a failure while 1 indicates success.
* Stopped using American spellings for function and variable names
* Can now limit number of titles returned when using find_by_title() (NOTE: IMDb will only return a maximum of 50 results)

## v0.1.0 - November 24, 2011
* **Initial release**
* Return movie data from IMDb ID
* Return movies matching title
* Anonymise requests to prevent IP address getting banned


tests/id.php
<?php
require '../class_IMDb.php';
$imdb = new IMDb(true);

$q = trim(stripslashes($_GET['q']));
?>
<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <title>Find by ID - IMDb-PHP-API</title>
</head>
<body>

<p>Enter a movie's IMDb ID to return its data. Example: <em>tt0068646</em></p>
<form action="" method="get">
<input type="text" name="q" maxlength="9" value="<?=$q?>" /> <input type="submit" value="Find" /><br />
<input id="nosummary" type="checkbox" name="nosummary"<? if($_GET['nosummary']=="on") print ' checked="checked"'?> /> <label for="nosummary">Do NOT summarise?</label>
</form>
<br />

<?php
if(!empty($q)){
	print '<h2>Results</h2>';
	if($_GET['nosummary']=="on") $imdb->summary=false;
	$movie = $imdb->find_by_id($q);

	print '<pre>';
	print_r($movie);
	print '</pre>';
}
?>

</body>
</html>

tests/title.php
<?php
require '../class_IMDb.php';
$imdb = new IMDb(true);

$q = trim(stripslashes($_GET['q']));
$year = $_GET['year'];
?>
<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <title>Find by Title - IMDb-PHP-API</title>
</head>
<body>

<p>Enter a search term to return data for matching movie titles. Example: <em>The Godfather</em></p>
<form action="" method="get">
<input type="text" name="q" value="<?=$q?>" /> <input type="submit" value="Search" /><br />
<input type="text" name="year" value="<?=$year?>" maxlength="4" /> Specify year (<em>Optional</em>)<br />
<input id="nosummary" type="checkbox" name="nosummary"<? if($_GET['nosummary']=="on") print ' checked="checked"'?> /> <label for="nosummary">Do NOT summarise?</label><br />
<input type="text" name="limit" value="<?=intval($_GET['limit'])?>" maxlength="2" /> Max. titles returned
</form>
<br />

<?php
if(!empty($q)){
	print '<h2>Results</h2>';
	if($_GET['nosummary']=="on") $imdb->summary=false;
	if(intval($_GET['limit'])>0) $imdb->titlesLimit = intval($_GET['limit']);
	$movies = $imdb->find_by_title($q, $year);

	print '<pre>';
	print_r($movies);
	print '</pre>';
}
?>

</body>
</html>

tests/charts.php
<?php
require '../class_IMDb.php';
$imdb = new IMDb(true);

$chart = trim(stripslashes($_GET['chart']));
$charts = array(
				"top" => "Top 250",
				"bottom" => "Bottom 100",
				"boxoffice" => "Box Office (US)"
          );
foreach($charts as $cid => $cname){
	if($chart==$cid) $s=' selected="selected"';
	$options .= '<option value="'.$cid.'"'.$s.'>'.$cname.'</option>'."\n";
	unset($s);
}
?>
<!doctype html>
<html>
<head>
  <meta charset="utf-8">
  <title>Charts - IMDb-PHP-API</title>
</head>
<body>

<p>Various charts from IMDb.</p>
<form action="" method="get">
<select name="chart">
<?=$options?>
</select>
<input type="submit" value="Show" />
</form>

<?php
print '<h2>Results</h2>';
if($chart=="top") $movies = $imdb->chart_top();
else if($chart=="bottom") $movies = $imdb->chart_bottom();
else if($chart=="boxoffice") $movies = $imdb->boxoffice();

print '<pre>';
if($movies) print_r($movies);
print '</pre>';
?>

</body>
</html>

class_IMDb.php
<?php
class IMDb
{
	
	private $baseurl = 'https://app.imdb.com/';
	private $params = array(
						'api'		=> 'v1',
						'appid'		=> 'iphone1_1',
						'apiPolicy'	=> 'app1_1',
						'apiKey'	=> '2wex6aeu6a8q9e49k7sfvufd6rhh0n',
						'locale'	=> 'en_US',
						'timestamp'	=> '0',
					  );
	private $anonymiser = 'http://anonymouse.org/cgi-bin/anon-www.cgi/';	// URL that will be prepended to the generated API URL.
	public $anonymise = false;
	public $summary = true;			// Set to true to return a summary of the film's details. Set to false to return everything.
	public $titlesLimit = 0;		// Limit the number of films returned by find_by_title() when summarised. 0 = unlimited (NOTE: IMDb returns a maximum of 50 results).
	
	// You can prevent certain types of titles being returned. IMDb classifies titles under one of the following categories:
	// feature, short, documentary, video, tv_series, tv_special, video_game
	// Titles whose category is in the $ignoreTypes array below will be removed from your results
	public $ignoreTypes = array('tv_series','tv_special','video_game');
	
	// By default, X rated titles and titles in the genre 'Adult' will also be ignored. Set to false to allow them.
	public $ignoreAdult = true;
	
	// Setting the following option to true will allow you to override any ignore options and force the title to be returned with find_by_id()
	public $forceReturn = false;
			
	function __construct($anonymise=false, $summary=true, $titlesLimit=0){
		$this->anonymise = $anonymise;		// should we anonymise requests?
		if(!$summary) $this->summary=false;	// overriding the default?
		if(intval($titlesLimit)>0)	$this->titlesLimit = intval($titlesLimit);		// Set titles limit if required
	}
	
	// Build URL based on the given parameters
	function build_url($method, $query="", $parameter=""){
		
		// Set timestamp parameter to current time
		$this->params['timestamp'] = $_SERVER['REQUEST_TIME'];
		
		// Build the URL and append query if we have one
		$unsignedUrl = $this->baseurl.$method.'?'.http_build_query($this->params);
		if(!empty($parameter) AND !empty($query)) $unsignedUrl .= '&'.$parameter.'='.urlencode($query);
		
		// Generate a signature and append to unsignedUrl to sign it.
		$sig = hash_hmac('sha1', $unsignedUrl, $this->params['apiKey']);
		$signedUrl = $unsignedUrl.'&sig=app1-'.$sig;
		
		// Anonymise the request?
		$signedUrl = $this->anonymise ? $this->anonymiser.$signedUrl : $signedUrl;
		
		return $signedUrl;
	}
	
	// Search IMDb by ID of film
	function find_by_id($id){
		if(strpos($id, "tt")!==0) $id = "tt".$id;
		$requestURL = $this->build_url('title/maindetails', $id, 'tconst');
		$json = $this->fetchJSON($requestURL);

		if(is_object($json->error)){
			$data = $this->errorResponse($json->error);
		}
		else{
			$data = $this->summary ? $this->summarise($json->data) : $json->data;
		}
		
		return $data;
	}
	
	// Search IMDb by title of film
	function find_by_title($title, $year=0){
		$requestURL = $this->build_url('find', $title, 'q');		
		$json = $this->fetchJSON($requestURL);
		
		// We'll usually have several "lists" returned in the JSON. Combine all these into one array.
		if(empty($json->data->results)){
			// IMDb doesn't return a proper error response in the event of 0 results being returned
			// so set our own failure message.
			$error->message = "No results found.";
			$matches = $this->errorResponse($error, true);
		}
		else{
			$results = $json->data->results;
			$matches = array();
			
			if($this->summary){
				$matches = $this->summarise_titles($results, intval($year));
			}
			else{
				for($i=0; $i<count($results); $i++){
					$matches = array_merge($matches, $results[$i]->list);
				}
			}
		}
		
		return $matches;
	}
	
	// Search IMDb by ID of person
	function person_by_id($id){
		if(strpos($id, "nm")!==0) $id = "nm".$id;
		$requestURL = $this->build_url('name/maindetails', $id, 'nconst');
		$json = $this->fetchJSON($requestURL);

		if(is_object($json->error)){
			$data = $this->errorResponse($json->error);
		}
		else{
			$data = $this->summary ? $this->summarise_person($json->data) : $json->data;
		}
		
		return $data;
	}
	
	// Search IMDb by name of person
	function person_by_name($name){
		$requestURL = $this->build_url('find', $name, 'q');		
		$json = $this->fetchJSON($requestURL);
		
		// We'll usually have several "lists" returned in the JSON. Combine all these into one array.
		if(empty($json->data->results)){
			// IMDb doesn't return a proper error response in the event of 0 results being returned
			// so set our own failure message.
			$error->message = "No results found.";
			$matches = $this->errorResponse($error, true);
		}
		else{
			$results = $json->data->results;
			$matches = array();
			
			if($this->summary){
				$matches = $this->summarise_people($results);
			}
			else{
				for($i=0; $i<count($results); $i++){
					$matches = array_merge($matches, $results[$i]->list);
				}
			}
		}
		
		return $matches;
	}
	
	// Top 250 Chart
	function chart_top(){
		$requestURL = $this->build_url('chart/top');
		$json = $this->fetchJSON($requestURL);

		if(is_object($json->error)){
			$data = $this->errorResponse($json->error);
		}
		else{
			$data = $json->data->list->list;
		}
		
		return $data;
	}
	
	// Bottom 100 Chart
	function chart_bottom(){
		$requestURL = $this->build_url('chart/bottom');
		$json = $this->fetchJSON($requestURL);

		if(is_object($json->error)){
			$data = $this->errorResponse($json->error);
		}
		else{
			$data = $json->data->list->list;
		}
		
		return $data;
	}
	
	// Box Office (US)
	function boxoffice(){
		$requestURL = $this->build_url('boxoffice');
		$json = $this->fetchJSON($requestURL);

		if(is_object($json->error)){
			$data = $this->errorResponse($json->error);
		}
		else{
			$data = $json->data;
		}
		
		return $data;
	}
	
	// Summarise - only return the most pertinent data (when returning data from IMDb ID)
	function summarise($obj){
	
		// If this is not an ignored type...
		if(!$this->is_ignored($obj->type, $obj->certificate->certificate, $obj->genres)){
			// ID with and without 'tt' prefix
			$s->id = substr($obj->tconst, 2);
			$s->tconst = $obj->tconst;
			
			// Title
			$s->title = $obj->title;
			
			// Year
			$s->year = $obj->year;
			
			// Plot and Tagline
			$s->plot = $obj->plot->outline;
			$s->tagline = $obj->tagline;
			
			// Votes + Rating
			$s->rating = $obj->rating;
			$s->votes = $obj->num_votes;
			
			// Genres
			if(is_array($obj->genres)){
				$s->genre = implode(", ", $obj->genres);
				$s->genres = $obj->genres;
			}else{
				$s->genre = "";
				$s->genres = "";
			}
			
			// Writers
			if(is_array($obj->writers_summary)){
				$i=0;
				foreach($obj->writers_summary as $writers){
					$writer[$i] = $writers->name->name;
					$s->writers_summary[$i]['nconst'] = $writers->name->nconst;
					$s->writers_summary[$i]['name'] = $writers->name->name;
					$s->writers_summary[$i]['attr'] = $writers->attr;
					$i++;
				}
				$s->writer = implode(", ", $writer);
			}else{
				$s->writer = "";
				$s->writers_summary = "";
			}
			
			// Directors
			if(is_array($obj->directors_summary)){
				$i=0;
				foreach($obj->directors_summary as $directors){
					$director[] = $directors->name->name;
					$s->directors_summary[$i]['nconst'] = $directors->name->nconst;
					$s->directors_summary[$i]['name'] = $directors->name->name;
					$i++;
				}
				$s->director = implode(", ", $director);
			}else{
				$s->director = "";
				$s->directors_summary = "";
			}
			
			// Cast
			if(is_array($obj->cast_summary)){
				$i=0;
				foreach($obj->cast_summary as $cast){
					$actor[] = $cast->name->name;
					$s->cast_summary[$i]['nconst'] = $cast->name->nconst;
					$s->cast_summary[$i]['name'] = $cast->name->name;
					$s->cast_summary[$i]['char'] = $cast->char;
					$i++;
				}
				$s->actors = implode(", ", $actor);
			}else{
				$s->actors = "";
				$s->cast_summary = "";
			}
			
			// Shorthand release date in the format of 'd MMM YYYY' and datestamp
			$s->released = !empty($obj->release_date->normal) ? date('j M Y', strtotime($obj->release_date->normal)) : "";
			$s->release_datestamp = $obj->release_date->normal;
			
			// Runtime
			$s->runtime = round($obj->runtime->time/60);
			
			// Certificate
			$s->certificate = $obj->certificate->certificate;
			
			// Poster
			$s->poster = $obj->image->url;
			
			// Type
			$s->type = $obj->type;
			
			// Response messages
			$s->response = 1;
			$s->response_msg = "Success";
		}
		else{
			// Response messages
			$s->response = 0;
			$s->response_msg = "Fail";
			$s->message = "Film type is ignored.";
		}
		
		return $s;
	}
	
	// Summarise - only return the most pertinent data (when returning multiple title data)
	function summarise_titles($objs, $yearMatch=0){
		
		$t=0;
				
		for($i=0; $i<count($objs); $i++){
			$list = $objs[$i]->list;
			
			// In each "list" of results we only want to return titles so ignore other results such as actors, characters etc.
			foreach($list as $obj){
				if(!empty($obj->tconst) AND !$this->is_ignored($obj->type) AND $this->yearMatch($obj->year, $yearMatch)){
					// ID with and without 'tt' prefix
					$s[$t]->id = substr($obj->tconst, 2);
					$s[$t]->tconst = $obj->tconst;
					
					// Title
					$s[$t]->title = $obj->title;
					
					// Year
					$s[$t]->year = $obj->year;
					
					// Comma-seperated list of actors
					$actor = array();
					if(is_array($obj->principals)){
						foreach($obj->principals as $cast){ $actor[] = $cast->name; }
						$s[$t]->actors = implode(", ", $actor);
					}else{
						$s[$t]->actors = "";
					}
					
					// Poster
					$s[$t]->poster = $obj->image->url;
					
					// Type
					$s[$t]->type = $obj->type;
					
					$t++;
					
					// Reached limit of titles?
					if($t==$this->titlesLimit) break 2;
				}
			}
		}
		
		// Response messages
		if($t>0){
			$s['response'] = 1;
			$s['response_msg'] = "Success";
		}
		else{
			$s['response'] = 0;
			$s['response_msg'] = "Fail";
		}
		
		return $s;
	}
	
	// Summarise Person - only return the most pertinent data (when returning data from IMDb ID)
	function summarise_person($obj){
	
		// ID with and without 'tt' prefix
		$s->id = substr($obj->nconst, 2);
		$s->nconst = $obj->nconst;
			
		// Name
		$s->name = $obj->name;
		$s->real_name = $obj->real_name;
			
		// Biography
		$s->bio = $obj->bio;
			
		// Birth and Death
		$s->birthday = !empty($obj->birth->date->normal) ? date('j M Y', strtotime($obj->birth->date->normal)) : "";
		$s->birthday_datestamp = $obj->birth->date->normal;
		$s->birthplace = $obj->birth->place;
		$s->deathday = !empty($obj->death->date->normal) ? date('j M Y', strtotime($obj->death->date->normal)) : "";
		$s->deathday_datestamp = $obj->death->date->normal;
		$s->deathplace = $obj->death->place;
		
		// Known For
		$s->known_for = $obj->known_for;
					
		// Image
		$s->image = $obj->image->url;
		$s->photos = $obj->photos;
			
		// Response messages
		$s->response = 1;
		$s->response_msg = "Success";
		
		return $s;
	}
	
	// Summarise - only return the most pertinent data (when returning multiple title data)
	function summarise_people($objs){
		
		$t=0;
				
		for($i=0; $i<count($objs); $i++){
			$list = $objs[$i]->list;
			
			// In each "list" of results we only want to return titles so ignore other results such as actors, characters etc.
			foreach($list as $obj){
				if(!empty($obj->nconst)){
					// ID with and without 'tt' prefix
					$s[$t]->id = substr($obj->nconst, 2);
					$s[$t]->nconst = $obj->nconst;
					
					// Name
					$s[$t]->name = $obj->name;
					
					// Known For
					$s[$t]->known_for = $obj->known_for;
					
					// Image
					$s[$t]->image = $obj->image->url;
					
					$t++;
					
					// Reached limit of titles?
					if($t==$this->titlesLimit) break 2;
				}
			}
		}
		
		// Response messages
		if($t>0){
			$s['response'] = 1;
			$s['response_msg'] = "Success";
		}
		else{
			$s['response'] = 0;
			$s['response_msg'] = "Fail";
		}
		
		return $s;
	}
	
	// Check if a title should be ignored. Returns true for yes, false for no.
	function is_ignored($type, $cert="", $genre=array()){
		if($this->forceReturn) return false;
		if(in_array($type, $this->ignoreTypes)) return true;
		if($this->ignoreAdult AND ($cert=="X" OR (!empty($genre) AND in_array("Adult",$genre)))) return true;
		
		return false;
	}
	
	// Check year matches
	function yearMatch($year, $yearMatch){
		if($yearMatch===0) return true;
		if($yearMatch==$year) return true;
		return false;
	}
	
	// Basic error handling
	function errorResponse($obj, $returnArray=false){
		$s->status = $obj->status;
		$s->code = $obj->code;
		$s->message = $obj->message;
		$s->response = 0;
		$s->response_msg = "Fail";
		
		if($returnArray) return (array)$s;
		else return $s;
	}
	
	
	// Perform CURL request on the API URL to fetch the JSON data
	function fetchJSON($apiUrl){
		$ch = curl_init($apiUrl);
		$headers[] = 'Connection: Keep-Alive'; 
		$headers[] = 'Content-type: text/plain;charset=UTF-8'; 
		curl_setopt($ch, CURLOPT_HTTPHEADER, $headers); 
		curl_setopt($ch, CURLOPT_HEADER, 0); 
		curl_setopt($ch, CURLOPT_TIMEOUT, 0); 
		curl_setopt($ch, CURLOPT_ENCODING , 'deflate');
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1); 
		curl_setopt($ch, CURLOPT_VERBOSE, 1); 
		$json = curl_exec($ch);
		$curl_errno = curl_errno($ch);
		$curl_error = curl_error($ch);
		curl_close($ch);
		
		// Errors?
		if ($curl_errno > 0){
			$data->error->message = 'cURL Error '.$curl_errno.': '.$curl_error;
		}
		else{	        
			// Decode the JSON response
			$data = json_decode($json);
		}
		
		return $data;
	}

}
?>